examples/vae_text/vae_train.py

```diff
@@ -196,111 +193,82 @@ def forward(self, data_batch: Dict, kl_weight: float) -> Dict:
+                batch_size = text_ids.size(0)
+                max_seq_len = text_ids.size(1) - 1
+                batch_max_seq_len = text_ids.new_full(
+                    (batch_size,), max_seq_len, dtype=torch.long)
+                output_p_embed = self.decoder_p_embedder(
+                    sequence_length=batch_max_seq_len)
+                output_w_embed = self.decoder_w_embedder(text_ids[:, :-1])
+                output_w_embed = output_w_embed * \
+                    self._config.hidden_size ** 0.5
+                output_embed = output_w_embed + output_p_embed
             else:
-                def _embedding_fn(ids, times):
-                    w_embed = self.decoder_w_embedder(ids)
-                    p_embed = self.decoder_p_embedder(times)
-                    return w_embed * config.hidden_size ** 0.5 + p_embed
-
-                outputs, _ = self.decoder(
-                    memory=dcdr_states,
-                    memory_sequence_length=torch.ones(dcdr_states.size(0)),
-                    max_decoding_length=100,
-                    decoding_strategy='infer_sample',
-                    embedding=_embedding_fn,
-                    start_tokens=start_tokens,
-                    end_token=end_token)
+                output_embed = None
+            outputs = self.decoder(
+                inputs=output_embed,
+                memory=decoder_states,
+                memory_sequence_length=torch.ones(decoder_states.size(0)),
+                helper=helper,
+                max_decoding_length=max_decoding_length)
         return outputs
 
 
 class VAEData(tx.data.MonoTextData):
     r""" Wrap `tx.data.MonoTextData` to implement data preprocessing.
     """
-    def collate(self, examples: List[List[str]]) -> Batch:
-        text_ids = [self._vocab.map_tokens_to_ids_py(sent) for sent in examples]
-        text_ids, lengths = tx.data.padded_batch(text_ids, self._pad_length,
-                                         pad_value=self._vocab.pad_token_id)
-        # Also pad the examples
-        pad_length = self._pad_length or max(lengths)
-        examples = [
-            sent + [''] * (pad_length - len(sent))
-            if len(sent) < pad_length else sent
-            for sent in examples
-        ]
-
-        text_ids = torch.from_numpy(text_ids).to(device=self.device)
-        lengths = torch.tensor(lengths, dtype=torch.long, device=self.device)
-        tmp_batch = []
-        for i, data in enumerate(text_ids):
-            filter_data = data[data != BLANK_ID]
-            delta = data.size(0) - filter_data.size(0)
-            tmp_batch.append(filter_data)
-            lengths[i] -= delta
-        text_ids = torch.stack(tmp_batch).to(device)
-        batch = {self.text_id_name: text_ids,
-                 self.length_name: lengths}
-        return Batch(len(examples), batch=batch)
+    def process(self, raw_example: str) -> List[str]:
+        # Truncates sentences and appends BOS/EOS tokens.
```

Comment:
I'm confused. Isn't this method exactly the same as `MonoTextData.process`? What is this specific preprocessing step that you require for VAE?
And also, if you still need the preprocessing steps of `MonoTextData`, you can always call `super().process(raw_example)` in this method.

And by the way, after the latest PR is merged, the `raw_example` argument in `MonoTextData.process` will become the tokenized input, i.e. a list of `str`s. It shouldn't require much change on your side though.

Reply:
Got it.